server/src/main/java/org/elasticsearch/search/internal/CancellableScorer.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.internal;
+
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.Scorer;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * A wrapper around scorer that executes checkCancelled on each document access.
+ *
+ * The main purpose of this class is to allow cancellation of search requests. Note that this class doesn't wrap bulk scorer, for that
+ * use {@link CancellableBulkScorer} instead.
+ */
+public class CancellableScorer extends Scorer {
```

Comment:
You could extend `FilterScorer` here which would simplify things a bit

Reply:
Yeah, that was my first attempt, but it wasn't possible due to [some methods being final](https://github.com/rsullivan00/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/search/FilterScorer.java#L73). But it is a really sensible suggestion, so I think I should probably add a comment explaining all this.